netket/nn/linear.py

```diff
@@ -330,10 +332,14 @@ def create_DenseSymm(
 
     See :ref:`netket.nn.DenseSymm` for the remaining parameters.
     """
-    if isinstance(permutations, Callable):
-        perm_fn = permutations
+    if isinstance(permutations, SymmGroup):
+        perms = permutations.indices()
+        perm_fn = lambda: perms
```

Comment:
I think ideally we could make this structure utterable, convertible to an array and hashable so we could pass it straight to Jax

Reply:
`SymmGroup` is now iterable and hashable. Also, it implements `__array__` (and `shape`) now, so it can be converted to a numpy array via
```python
>>> np.asarray(g.translations())
array([[0, 1, 2, 3],
       [3, 0, 1, 2],
       [2, 3, 0, 1],
       [1, 2, 3, 0]])
```
instead of `indices()`.

I'm still not sure we should use it in `DenseSymm` directly, but we could do it now.

Reply:
What makes you not sure?
I think it's a nice way to solve our issue of the constructor...

Reply:
>  I think it's a nice way to solve our issue of the constructor...

That's what I'm unsure about. Say we already have our permutations as an array -- then we still cannot pass it to `DenseSymm` directly, neither could we pass `AbstractGraph`, if it requires a (hashable) `SymmGroup` now (I think @gcarleo's question was related to this).

That's essentially the same situation we have now just with a different trade-off. Thus, I think this issue is orthogonal to this PR and I don't think forcing `SymmGroup` into the role of a hashable array wrapper just because we need this somewhere else is necessarily a good idea. (Maybe we should just write an immutable `HashableArray` class for this purpose and/or suggest adding init-only keywords to `flax`; I think both would be better solutions than forcing everything to be a `SymmGroup`.)

Reply:
@PhilipVinc 

> (Maybe we should just write an immutable HashableArray class for this [...])

By the way: I've implemented this and would like to make a PR (once this is merged to avoid conflicts):

https://github.com/femtobit/netket/compare/symmetry-groups...femtobit:hashable-array

I think, while slightly anoying, having a `HashableArray` wrapper is the best solution currently: Even if we work around it here with `SymmGroup`, the need for array-valued model attributes will probably come up again with other values. Having a class for this makes the issue and solution explicit (and can also be used by users who define their own models).